tapdb/supply_commit.go

```diff
@@ -0,0 +1,1303 @@
+		return e.Encode(w)
+
+	default:
+		return fmt.Errorf("unknown event type: %T", event)
+	}
+}
+
+// deserializeSupplyUpdateEvent decodes bytes into a SupplyUpdateEvent.
+func deserializeSupplyUpdateEvent(typeName string,
+	r io.Reader) (supplycommit.SupplyUpdateEvent, error) {
+
+	switch typeName {
+	case "mint":
+		mint := new(supplycommit.NewMintEvent)
+		if err := mint.Decode(r); err != nil {
+			return nil, fmt.Errorf("failed to decode mint "+
+				"event: %w", err)
+		}
+
+		return mint, nil
+	case "burn":
+		burn := new(supplycommit.NewBurnEvent)
+		if err := burn.Decode(r); err != nil {
+			return nil, fmt.Errorf("failed to decode burn "+
+				"event: %w", err)
+		}
+
+		return burn, nil
+	case "ignore":
+		ignore := new(supplycommit.NewIgnoreEvent)
+		if err := ignore.Decode(r); err != nil {
+			return nil, fmt.Errorf("failed to decode ignore "+
+				"event: %w", err)
+		}
+
+		return ignore, nil
+
+	default:
+		return nil, fmt.Errorf("unknown update type name: %s", typeName)
+	}
+}
+
+// InsertPendingUpdate attempts to insert a new pending update into the
+// update log of the target supply commit state machine.
+func (s *SupplyCommitMachine) InsertPendingUpdate(ctx context.Context,
+	assetSpec asset.Specifier, event supplycommit.SupplyUpdateEvent) error {
+
+	groupKey := assetSpec.UnwrapGroupKeyToPtr()
+	if groupKey == nil {
+		return ErrMissingGroupKey
+	}
+	groupKeyBytes := groupKey.SerializeCompressed()
+
+	var writeTx SupplyCommitTxOptions // Default read-write
+	return s.db.ExecTx(ctx, &writeTx, func(db SupplyCommitStore) error {
+		// First, we'll upsert a new state machine. We pass a null state
+		// name, as it'll be made with default if doesn't exist. The
+		// query returns the actual state ID set and the latest
+		// commitment ID.
+		upsertResult, err := db.UpsertSupplyCommitStateMachine(
+			ctx, SupplyCommitMachineParams{
+				GroupKey:  groupKeyBytes,
+				StateName: sql.NullString{},
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("failed initial upsert for "+
+				"state machine: %w", err)
+		}
+		currentStateID := upsertResult.CurrentStateID
+
+		// Make sure that we're in either the DefaultState or
+		// UpdatesPendingState state as a sanity check.
+		currentState, err := intToState(currentStateID)
+		if err != nil {
+			// This indicates an unexpected state ID returned
+			// from the DB.
+			return fmt.Errorf("invalid state ID %d "+
+				"returned from upsert: %w", currentStateID, err)
+		}
+		currentStateName := currentState.String()
+		if currentStateName != "DefaultState" &&
+			currentStateName != "UpdatesPendingState" {
+
+			return fmt.Errorf("cannot insert pending "+
+				"update in state: %s", currentStateName)
+		}
+
+		// Now that we know the state machine is in the proper state,
+		// we'll fetch the transition ID, which will be needed below.
+		var transitionID int64
+		existingTransitionID, err := db.QueryExistingPendingTransition(
+			ctx, groupKeyBytes,
+		)
+		if err != nil {
```

Comment:
Feels like this whole block could be simplified with a `switch`, instead of the `else` cases.

Reply:
Nice suggestion, was also able to get rid of some of the linter overrides as we've eliminated a layer of nesting.